src/bgg/circuit/mod.rs

```diff
@@ -192,60 +196,131 @@ impl PolyCircuit {
         order
     }
 
+    /// Computes a levelized grouping of gate ids.
+    /// Input wires (keys 0..=num_input) are assigned level 0.
+    /// Each non‐input gate's level is defined as max(levels of its inputs) + 1.
+    fn compute_levels(&self) -> Vec<Vec<usize>> {
+        let mut gate_levels: HashMap<usize, usize> = HashMap::new();
+        let mut levels: Vec<Vec<usize>> = vec![vec![]];
+        for i in 0..=self.num_input {
+            gate_levels.insert(i, 0);
+        }
+        let orders = self.topological_order();
+        for gate_id in orders.into_iter() {
+            let gate = self.gates.get(&gate_id).expect("gate not found");
+            if gate.input_gates.len() == 0 {
+                gate_levels.insert(gate_id, 0);
+                levels[0].push(gate_id);
+                continue;
+            }
+            let level = gate
+                .input_gates
+                .iter()
+                .map(|id| *gate_levels.get(id).expect("input gate not found"))
+                .max()
+                .expect("max level not found") +
+                1;
+            gate_levels.insert(gate_id, level);
+            if levels.len() <= level {
+                levels.resize(level + 1, vec![]);
+            }
+            levels[level].push(gate_id);
+        }
+        levels
+    }
+
     /// Evaluate the circuit using an iterative approach over a precomputed topological order.
     pub fn eval<E: Evaluable>(&self, params: &E::Params, one: &E, inputs: &[E]) -> Vec<E> {
         #[cfg(debug_assertions)]
         {
             assert_eq!(self.num_input(), inputs.len());
             assert_ne!(self.num_output(), 0);
         }
-        let num_gates = self.gates.len();
-        let mut wires: Vec<Option<E>> = vec![None; num_gates];
-        wires[0] = Some(one.clone());
+
+        let wires = DashMap::new();
+        let levels = self.compute_levels();
+        debug_mem(format!("Levels: {:?}", levels));
+        debug_mem("Levels are computed");
```

Comment:
nit but no need?

Reply:
I think it's fine to keep these logs for additional information.